### PR TITLE
Changed masked out rows logsumexp to be -inf and not zero

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134650

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by yanboliang — APPROVED** [review]

**Review by BoyuanFeng — APPROVED** [review]

File: torch/_inductor/kernel/flex_decoding.py [hunk 1]
```diff
@@ -546,8 +546,8 @@ def create_flex_decoding_kernel(*args, **kwargs):
     # See [Note] Handle fully masked out rows:
     # g_M Is the global max among split kv blocks.
     masked_rows = lowerings[aten.eq](g_M, -float("inf"))
-    g_M = lowerings[aten.where](masked_rows, 0.0, g_M)
     adj_M = lowerings[aten.sub](buf_M, g_M)
+    adj_M = lowerings[aten.where](masked_rows, 0, adj_M)
```
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_inductor/kernel/flex_decoding.py:550]: The previous order was actually incorrect for the semantics where `logsumexp` of a masked out row is `-inf`. Specifically, the previous masking was done because we're computing the "difference in max values" so we can rescale it appropriately, and if `g_M` and `buf_M` were both `-inf` this would cause nans (i.e. `-inf - (-inf) => nan`).

However, we then compute `log(L) + M`. If `M` was masked to 0 here, then the end result would be 0, which is not the `logsumexp` value we want.

File: test/inductor/test_flex_attention.py [hunk 2]
```diff
@@ -1641,6 +1641,86 @@ def test_force_write_lse(self):
 
         torch.testing.assert_close(lse_eager, lse_compiled, atol=3e-3, rtol=0)
 
+    @supported_platform
+    @common_utils.parametrize("backend", ["flex_attention", "flex_decode", "eager"])
+    def test_lse_masked_output(self, backend):
+        if backend == "flex_decode":
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:1647]: BE: we really need to unify our FlexDecoding and FlexAttention testing

File: test/inductor/test_flex_attention.py [hunk 3]
```diff
@@ -1641,6 +1641,86 @@ def test_force_write_lse(self):
...
+            return_lse=True,
+            kernel_options=kernel_options,
+        )
+        q, k, v = make_tensor(), make_tensor(), make_tensor()
+        gradOut = make_tensor(requires_grad=False)
+
+        x_local, lse_local = local_attn(q, k, v)
+        x_global, lse_global = global_attn(q, k, v)
+
+        max_lse = torch.maximum(lse_local, lse_global)
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:1702]: BE: we also should add a public version of this: https://github.com/pytorch/pytorch/blob/092349dcddb39b9844d730bb0da883921e45d3e7/torch/distributed/_tensor/experimental/attention.py#L79

mabye just functional though

**Review by drisspg — APPROVED** [review]
